src/System.Text.Json/src/System/Text/Json/JsonElement.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            if (TryGetValue(out double value))
+            {
+                return value;
+            }
+
+            throw new FormatException();
+        }
+
+        public ChildEnumerator EnumerateChildren()
+        {
+            if (_parent == null)
+            {
+                throw new InvalidOperationException();
+            }
+
+            return new ChildEnumerator(this);
+        }
+
+        public override string ToString()
+        {
+            switch (Type)
+            {
+                case JsonTokenType.None:
+                case JsonTokenType.Null:
+                    return string.Empty;
+                case JsonTokenType.True:
+                    return bool.TrueString;
+                case JsonTokenType.False:
+                    return bool.FalseString;
+                case JsonTokenType.Number:
+                case JsonTokenType.StartArray:
+                case JsonTokenType.StartObject:
+                case JsonTokenType.Comment:
+                {
+                    // null parent should have hit the None case
+                    Debug.Assert(_parent != null);
+                    return _parent.GetRawValueAsString(_idx);
+                }
+                case JsonTokenType.PropertyName:
+                case JsonTokenType.String:
+                    return GetString();
+                case JsonTokenType.EndArray:
+                case JsonTokenType.EndObject:
+                default:
+                    Debug.Fail($"No handler for {nameof(JsonTokenType)}.{Type}");
+                    return string.Empty;
+            }
+        }
+
+        public struct ChildEnumerator
+        {
+            private JsonElement _target;
+            public JsonElement Current { get; private set; }
+            private readonly int _endIdx;
+
+            internal ChildEnumerator(JsonElement target)
+            {
+                _target = target;
+                Current = default;
+                _endIdx = _target._parent.GetEndIndex(_target._idx, includeEndElement: false);
+            }
+
+            public ChildEnumerator GetEnumerator()
+            {
+                ChildEnumerator ator = this;
+                ator.Current = default;
+                return ator;
+            }
+
+            public bool MoveNext()
+            {
+                if (Current._idx >= _endIdx)
+                    return false;
+
+                int nextIdx;
+
+                if (Current._parent == null)
+                {
+                    nextIdx = _target._idx + JsonDocument.DbRow.Size;
+                }
+                else
+                {
+                    nextIdx = _target._parent.GetEndIndex(Current._idx, includeEndElement: true);
+                }
+
+                if (nextIdx >= _endIdx)
+                {
+                    Current = default;
+                    return false;
+                }
+
+                Current = new JsonElement(_target._parent, nextIdx);
```

Comment:
Is this approach more efficient than having Current manufacture a JsonElement in its getter?  Presumably in the 99.9% case Current is only accessed once, and storing Current means the enumerator is larger than it needs to be.